src/components/PageSearchComponent.tsx

```diff
@@ -0,0 +1,133 @@
+import { Input } from "@/components/ui/input"
+import { useAxios } from "@/hooks/use-axios"
+import { useLocation } from "wouter"
+import React, { useState, useEffect } from "react"
+import { useQuery } from "react-query"
+import { Alert } from "./ui/alert"
+import { useSnippetsBaseApiUrl } from "@/hooks/use-snippets-base-api-url"
+import { Loader2, Search } from "lucide-react"
+import { SnippetCard } from "./SnippetCard"
+
+interface PageSearchComponentProps {
+  onResultsFetched?: (results: any[]) => void
+}
+
+const PageSearchComponent: React.FC<PageSearchComponentProps> = ({
+  onResultsFetched,
+}) => {
+  const [location, setLocation] = useLocation()
+  const axios = useAxios()
+  const snippetsBaseApiUrl = useSnippetsBaseApiUrl()
+
+  const initialQuery =
+    new URLSearchParams(window.location.search).get("q") || ""
+  const [searchQuery, setSearchQuery] = useState(initialQuery)
+
+  const { data: searchResults, isLoading } = useQuery(
+    ["snippetSearch", searchQuery],
+    async () => {
+      if (!searchQuery) return []
+      const { data } = await axios.get("/snippets/search", {
+        params: { q: searchQuery },
+      })
+      if (onResultsFetched) {
+        onResultsFetched(data.snippets)
+      }
+      return data.snippets
+    },
+    { enabled: Boolean(searchQuery) },
+  )
+
+  useEffect(() => {
+    const baseUrl = location.split("?")[0]
+    const newUrl = searchQuery
+      ? `${baseUrl}?q=${encodeURIComponent(searchQuery)}`
+      : baseUrl
+    if (newUrl !== location) {
+      setLocation(newUrl)
+    }
+  }, [searchQuery, location, setLocation])
+
+  useEffect(() => {
+    const params = new URLSearchParams(window.location.search)
+    const urlQuery = params.get("q")
+    if (urlQuery !== null && urlQuery !== searchQuery) {
+      setSearchQuery(urlQuery)
+    }
+  }, [location])
```

Comment:
antipattern. You can just use `useState(params.get("q") ?? "")`

Remember, every useEffect triggers a render and makes the page slow

```suggestion
```